### PR TITLE
Implement tool data bundles

### PR DESCRIPTION
**Pull request by jmchilton**
- Create a serializable and typed and validated version of the ``data_manager_conf.xml`` files - called ``DataTableBundleProcessorDescription`` (because these files describe how to process data to modify data tables) and placed in ``galaxy.tool_util.data.bundles.models``. 
- Move code out of ``galaxy.tools`` and into ``galaxy.tool_util.data`` for taking a dataset containing the data manager JSON output and the above models and "apply" it to data tables - implement unit tests for the two major patterns I've found. As a result of moving this into ``galaxy-tool-util`` and cleaning up dependencies these unit tests are very clean and do not depend on app or config.
- Add a job creation (``POST /api/tools``) input parameter ``data_manager_mode`` that defaults to the current behavior of updating data tables with ``populate``. This parameter however can be overridden and set to ``dry_run`` which will run the data manager and then do nothing (data tables will not be updated) or ``bundle`` which will take the extra files path of the ``data_manager_json`` and add a JSON index file that turns it into a "tool data bundle". The index includes the JSON-ified version of the ``data_manager_conf.xml`` described above which is needed to turn the content of the extra files path into a data manager value.
- The data manager bundle can be downloaded as a zip file using basically just the normal dataset display method - though I did update the ``data_manager_json`` datatype to reflect we care about the composite contents - this helps force the download to be an achieve.
- Finally, to make the tool data bundles useful, this PR adds an admin-only ``POST /api/tool_data`` FastAPI route. This can take a dataset ID or a URI (the URI can reference an archive or a directory if it is a ``file://`` URI). The dataset or file or directory is assumed to a data manager bundle and the bundle will be imported into the Galaxy instance.

This includes lots of unit tests, new typing, and a couple key integration tests.

This should allow the IDC CI to run data managers on large public instances (.org and .eu) and pull down the results and apply them to CVMFS out of band.

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

File: lib/galaxy/webapps/galaxy/services/tools.py [hunk 1]
```diff
@@ -163,7 +163,8 @@ def _create(self, trans: ProvidesHistoryContext, payload, **kwd):
         )
 
         input_format = str(payload.get("input_format", "legacy"))
-
+        if "data_manager_mode" in payload:
```
Comment (mvdbeek) [inline lib/galaxy/webapps/galaxy/services/tools.py:166]: I guess we'll eventually want to document all those parameter, should we have a separate endpoint to run data manager tools to not clutter this ? Similar to how we deal with `fetch_data` ?

Comment (mvdbeek) [reply lib/galaxy/webapps/galaxy/services/tools.py:166]: Or perhaps add this as a parameter to each data manager tool ?

Comment (jmchilton) [reply lib/galaxy/webapps/galaxy/services/tools.py:166]: https://github.com/galaxyproject/galaxy/blob/0569742c60ac936f0248a445543138c5ac9bb952/lib/galaxy/webapps/galaxy/services/jobs.py#L56 ... tool state branch seems the right place to do that? Will you accept an IOU for that.

**Comment by wm75** [comment]
This is awesome @jmchilton!
The bundle idea and the metadata stored with it has a bit of overlap with what I've implemented for https://github.com/galaxyproject/tools-iuc/tree/master/data_managers/data_manager_packaged_annotation_data. The idea there is to share metadata yaml files for genome annotation data and have a DM that can read them and install the requested data from them. Not sure if there are useful additional ideas for your PR in there, but it may be worth taking a look at least?

**Comment by wm75** [comment]
Oh and https://github.com/galaxyproject/tools-iuc/tree/master/tools/packaged_annotation_loader is a tool demonstrating loading contents from the packaged_annotation_data data table.

File: lib/galaxy/webapps/galaxy/services/tools.py (hunk 1 above)
Comment (mvdbeek) [reply lib/galaxy/webapps/galaxy/services/tools.py:166]: Sure, sounds good!

**Comment by jmchilton** [comment]
@wm75  Very cool - I had not seen those and I'm glad I know they exist now. I looked those over last week and I was impressed with what you were able to within the data manager framework. I'm not sure if and how to blend it into this work though. I will keep it in mind going forward though.

**Comment by jmchilton** [comment]
Oh wow - that is a valuable test! Shed code digging deep into Galaxy's internals detected. 

```
Traceback (most recent call last):
  File "/home/runner/work/galaxy/galaxy/galaxy root/lib/galaxy/jobs/runners/__init__.py", line 262, in prepare_job
    job_wrapper.prepare()
  File "/home/runner/work/galaxy/galaxy/galaxy root/lib/galaxy/jobs/__init__.py", line 1247, in prepare
    tool_evaluator.set_compute_environment(compute_environment, get_special=get_special)
  File "/home/runner/work/galaxy/galaxy/galaxy root/lib/galaxy/tools/evaluation.py", line 162, in set_compute_environment
    self.execute_tool_hooks(inp_data=inp_data, out_data=out_data, incoming=incoming)
  File "/home/runner/work/galaxy/galaxy/galaxy root/lib/galaxy/tools/evaluation.py", line 170, in execute_tool_hooks
    "exec_before_job", self.app, inp_data=inp_data, out_data=out_data, tool=self.tool, param_dict=incoming
  File "/home/runner/work/galaxy/galaxy/galaxy root/lib/galaxy/tools/__init__.py", line 2168, in call_hook
    return code(*args, **kwargs)
  File "/tmp/tmpywg04oef/toolshed.g2.bx.psu.edu/repos/iuc/data_manager_manual/6a6a2e4cc222/data_manager_manual/data_manager/data_manager_manual.py", line 56, in exec_before_job
    data_table_filename = data_table.get_filename_for_source(data_manager, None)
  File "/home/runner/work/galaxy/galaxy/galaxy root/lib/galaxy/tool_util/data/__init__.py", line 617, in get_filename_for_source
    source_repo_info = source.dict()
AttributeError: Error in 'Manual Data Manager' hook 'exec_before_job', original message: 'DataManager' object has no attribute 'dict'
```

File: lib/galaxy/tool_util/data/bundles/models.py [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+    name: str
+    owner: str
+    installed_changeset_revision: str
+
+    class Config:
+        # we use dictionary equality (yuk) so definitely make sure this is okay.
+        extra = Extra.forbid
+
+
+class DataTableBundle(BaseModel):
```
Comment (mvdbeek) [inline lib/galaxy/tool_util/data/bundles/models.py:169]: Could / should this include a serialized version of the job that generated the bundle ?

**Review by mvdbeek — APPROVED** [review]

**Pull request merged**